[spec] add supplementary guidance around env context propagation (#4548)
open-telemetry/opentelemetry-specification: specification/context/env-carriers.md

- When spawning multiple child processes with different contexts or baggage,
  each child SHOULD receive its own copy of the environment variables with
  appropriate information.

> - The onus is on the application owner for receiving the set context from the

Reviewer: nit: maybe

```suggestion
- The onus is on the application developer for receiving the set context from the
```

?

Author: `application owner` comes from the glossary so i'd like to leave it at that https://opentelemetry.io/docs/specs/otel/glossary/#application-owner

>   SDK and passing it to it's own process spawning mechanism. The language

Reviewer: ```suggestion
  SDK and passing it to its own process spawning mechanism. The language
```

>   passing it to it's own process spawning mechanism. The SDK SHOULD NOT handle

Reviewer 2: What do we mean by "end-user" here? I find this term here very confusing.

Author: The application/service author who imports the SDK & API and instruments accordingly.

Reviewer 2: I think it should be "application owner" or "application developer".
See: https://opentelemetry.io/docs/specs/otel/glossary/#application-owner

CC @open-telemetry/docs-maintainers

Author: Thanks for the link. That sounds reasonable to me, I'll make the adjustment.

>   spawning processes for the end-user.

Reviewer 2: What do we mean by SDK here? There is no place in the SDK that spawns processes. Also propagators are not part of SDK.

>   implementations MUST NOT handle spawning processes.

Reviewer: what does it mean?  Implementations don't spawn processes, so there is no chance they could handle it, but instrumentation libraries should, in some specific cases, inject and extract context over env vars.

Author: they don't spawn process, nor should they add that functionality now that we have a way to propagate context through the environment. I spelled this out so that there's clear deliniation that the SDK won't handle process spawning. It'll only hand pulling context from the environment, and setting context into a dictionary/map that the user will use when spawning a process.

- For maximum compatibility, implementations MUST:
  - Use uppercase names consistently (`TRACEPARENT` not `TraceParent`).
  - Use the canonical case when setting environment variables.

## Supplementary Guidelines

> **IMPORTANT**: This section is non-normative and provides implementation
> guidance only. It does not add requirements to the specification.

> Language SDKs have flexibility in how they implement environment variable

Reviewer 2: SDK is part of language implementation of OpenTelemetry. Propagators are nor part of the SDK.

```suggestion
Language implementations of OpenTelemetry have flexibility in how they implement environment variable
```

context propagation. Two main approaches have been identified as viable.

### Approach 1: Providing a dedicated `EnvironmentContextPropagator`

SDKs can create a dedicated propagator for environment variables. For example,

> the [Swift SDK][swift] implements a custom `EnvironmentContextPropagator` that

Reviewer 2: ```suggestion
the [OTel Swift][swift] implements a custom `EnvironmentContextPropagator` that
```

handles the environment-specific logic internally, in essence decorating the
`TextMapPropagator`.

### Approach 2: Using the carriers directly through `Setters` and `Getters`

> SDKs can use the existing `TextMapPropagator` interface directly with

Reviewer 2: ```suggestion
Language implementations can use the existing `TextMapPropagator` interface directly with
```

environment-specific carriers. Go and Python SDKs follow this pattern by
providing:

- `EnvironmentGetter` - creates an in-memory copy of the current environment
  variables and reads context from that copy.
- `EnvironmentSetter` - writes context to a dictionary/map and provides the
  dictionary/map to the end-user for them to use when spawning processes.

### Common Behaviors

Both approaches achieve the same outcome while offering different developer

> experiences. SDK implementers may choose either approach based on their

Reviewer 2: ```suggestion
experiences. Language implementations may choose either approach based on their
```

- For maximum compatibility, implementations MUST:
  - Use uppercase names consistently (`TRACEPARENT` not `TraceParent`).
  - Use the canonical case when setting environment variables.

## Supplementary Guidelines

> **IMPORTANT**: This section is non-normative and provides implementation
> guidance only. It does not add requirements to the specification.

Language implementations of OpenTelemetry have flexibility in how they implement
environment variable context propagation. Two main approaches have been
identified as viable.

### Approach 1: Providing a dedicated `EnvironmentContextPropagator`

SDKs can create a dedicated propagator for environment variables. For example,
the [OTel Swift][swift] implements a custom `EnvironmentContextPropagator` that
handles the environment-specific logic internally, in essence decorating the
`TextMapPropagator`.

### Approach 2: Using the carriers directly through `Setters` and `Getters`

Language implementations can use the existing `TextMapPropagator` interface directly with

> environment-specific carriers. Go and Python SDKs follow this pattern by

Author: Ok, I've added both the python & go prototypes as examples.
